Add space.celestiaproject.Celestia

ColinKinloch opened the pull request:
### Please confirm your submission meets all the criteria

- [x] I have read the [App Requirements][reqs] and [App Maintenance][maint] pages.
- [ ] My pull request follows the instructions at [App Submission][submission].
- [x] I have [built][build] and tested the submission locally.
- [x] I am using only the minimal set of permissions. *(If not, please explain each non-standard permission.)*
- [x] All assets referenced in the manifest are redistributable by any party.  If not, the unredistributable parts are using an extra-data source type.
- [ ] I am an upstream contributor to the project. If not, I contacted upstream developers about submitting their software to Flathub. **Link:**
- [x] The domain used for the application ID is controlled by the application developers either directly or through the code hosting (e.g. GitHub, GitLab, SourceForge, etc.). The [application id guidelines][app-id] are followed.
- [x] Any additional patches or files have been submitted to the upstream projects concerned. *(If not, explain why.)*

* I've yet to seek the developers permission though they seem open to flatpak https://github.com/CelestiaProject/Celestia/issues/2049.
* This manifest uses cmake and the `master` branch. The current stable release uses autotools.
* This build includes 3 front ends / desktop files: Qt6, gtk3 and sdl. I think it would probably be fine with just Qt6.


[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[maint]: https://docs.flathub.org/docs/for-app-authors/maintanance
[submission]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission
[app-id]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id

hfiguiere commented on space.celestiaproject.Celestia.yml:
```diff
@@ -0,0 +1,88 @@
+    sources:
+      - type: git
+        url: https://github.com/CelestiaProject/CelestiaContent.git
+        branch: master
```
use of git branch isn't allowed. this must be a commit and/or tag.

hfiguiere replied:
this is why the build fails.

hfiguiere replied:
still why the build fail.

hfiguiere replied:
and it will fail again.

hfiguiere commented on space.celestiaproject.Celestia.metainfo.xml:
why not using https://github.com/CelestiaProject/Celestia/blob/master/src/celestia/gtk/data/space.celestia.celestia_gtk.metainfo.xml (maybe with patches) ?

hfiguiere replied:
or the one in qt6

ColinKinloch replied:
I should submit my changes upstream.

hfiguiere commented on space.celestiaproject.Celestia.yml:
```diff
@@ -0,0 +1,88 @@
+    autodelete: false
+
+cleanup-commands:
+  - mkdir -p /app/lib/ffmpeg
```
this doesn't belong to cleanup-commands. It belong to a `post-install`. usually the main module.

ColinKinloch commented on space.celestiaproject.Celestia.yml:
```diff
@@ -0,0 +1,88 @@
+        ICON_IN="./celestia-logo.png"
+        ICON_OUT="celestia.png"
+        SIZES='16 24 32 48 64 128 256'
+        for s in ${SIZES} ; do
```
I spotted this, it's not the current logo. But now that you mention it there's a 1024px version here:  https://celestiaproject.space/images/logo.png

hfiguiere replied:
maybe they should fix it. to me it looks like the thing to use. even the desktop file next to it say so.


Now if you have a SVG it's best. 1024x1024 isn't a valid size.

bbhtt commented on space.celestiaproject.Celestia.metainfo.xml:
```diff
@@ -0,0 +1,49 @@
+  <launchable type="desktop-id">space.celestiaproject.Celestia.desktop</launchable>
+  <url type="homepage">https://celestiaproject.space</url>
+
+  <content_rating type="oars-1.0">
+  </content_rating>
```
```suggestion
  <content_rating type="oars-1.1" />
```

https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/#oars-information

bbhtt commented on space.celestiaproject.Celestia.metainfo.xml:
```diff
@@ -0,0 +1,49 @@
+      <image>https://celestiaproject.space/images/gallery/proxima.png</image>
+    </screenshot>
+  </screenshots>
+</component>
```
There's no release info here https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/#release

bbhtt commented on space.celestiaproject.Celestia.yml:
```diff
@@ -0,0 +1,88 @@
+          install -p -Dm644 "${ICON_OUT}" -t "${FLATPAK_DEST}/share/icons/hicolor/${SIZE}/apps/"
+        done
+    sources:
+      - type: git
+        url: https://github.com/CelestiaProject/Celestia.git
+        branch: master
+        #commit: 315b2d798496898c69240ec68e2059bd7d8c246e
```
Should be from a tag, not master or at least a commit

ColinKinloch commented:
bot, build space.celestiaproject.Celestia

ColinKinloch commented on space.celestiaproject.Celestia.yml:
```diff
@@ -0,0 +1,86 @@
+    sources:
+      - type: git
+        url: https://github.com/CelestiaProject/Celestia.git
+        branch: master
```
Oh, I see.
Will it build if it's only a commit?
Or should I leave building it until there's a release tag?

hfiguiere replied:
a commit is fine. has always been.

ColinKinloch commented:
bot, build space.celestiaproject.Celestia

ColinKinloch commented:
bot, build space.celestiaproject.Celestia

ColinKinloch commented:
bot, build space.celestiaproject.Celestia

RacerBG commented:
Am I doing something wrong? I'm trying to install it with the command from above and I'm getting the following error:

`error: Can't load uri https://dl.flathub.org/build-repo/84053/space.celestiaproject.Celestia.flatpakref: Server returned status 404`

ColinKinloch commented:
It's a test build so it gets cleaned up after a certain amount of time.

RacerBG commented:
bot, build space.celestiaproject.Celestia

ColinKinloch commented:
I just gave it a test with `--nosocket=wayland`, it seems to work fine my end under XWayland.
Is there a way to get this application to print a log?

RacerBG commented:
> I just gave it a test with `--nosocket=wayland`, it seems to work fine my end under XWayland. Is there a way to get this application to print a log?

Yes, run it through the terminal:

`Detected locale "C" with character encoding "ANSI_X3.4-1968", which is not UTF-8.
Qt depends on a UTF-8 locale, and has switched to "C.UTF-8" instead.
If this causes problems, reconfigure your locale. See the locale(1) manual
for more information.
Qt: Session management error: Could not open network socket
terminate called after throwing an instance of 'std::runtime_error'
what():  locale::facet::_S_create_c_locale name not valid`

ColinKinloch commented:
Okay, I think I've reproduced your issue with `flatpak run --env=LC_ALL=C.ANSI_X3.4-1968 --nosocket=wayland --command=celestia-qt6 space.celestiaproject.Celestia//test`

So could you test it with:
```
flatpak run --env=LC_ALL=C.UTF-8 --command=celestia-sdl space.celestiaproject.Celestia//test
```

Testing it without the x11 and wayland changes in flatseal would also be useful.

ColinKinloch commented:
Could you give me the output of `localectl` also?

RacerBG commented:
> Could you give me the output of `localectl` also?

It works now! The 3 backends are starting just fine, the SDL one doesn't have a menu bar though, I don't know if this is intentional.

The output of localectl is:
`System Locale: LANG=bg_BG.UTF-8
LANGUAGE=en_US:en
LC_MESSAGES=en_US.UTF-8
VC Keymap: (unset)
X11 Layout: us,bg
X11 Model: pc105
X11 Variant: ,phonetic
X11 Options: grp:alt_shift_toggle,grp_led:scroll`

locale -a:
`bg_BG.utf8
C
C.utf8
en_US.utf8
POSIX`

locale:
`LANG=en_US.UTF-8
LANGUAGE=en_US:en
LC_CTYPE=bg_BG.UTF-8
LC_NUMERIC=bg_BG.UTF-8
LC_TIME=bg_BG.UTF-8
LC_COLLATE=bg_BG.UTF-8
LC_MONETARY=bg_BG.UTF-8
LC_MESSAGES=en_US.UTF-8
LC_PAPER=bg_BG.UTF-8
LC_NAME=bg_BG.UTF-8
LC_ADDRESS=bg_BG.UTF-8
LC_TELEPHONE=bg_BG.UTF-8
LC_MEASUREMENT=bg_BG.UTF-8
LC_IDENTIFICATION=bg_BG.UTF-8
LC_ALL=`

I will test it in depth now to see if everything works. For some reason the GTK version looks the best while the Qt6 version does not respect the system Qt looks.

RacerBG commented:
First thoughts: give it more permissions: xdg-pictures:create and xdg-videos:create

Celestia have capture movie and capture image functions so let's take advantage of that. :smiley: 

A question: Where am I supposed to put the addons? For example textures, spacecrafts, star catalogues?

Overall it works fine for me on X11 but you should find a fix for the locale or just use `--env=LC_ALL=C` by default **if nothing else helps**. I rely entirely on flatpaks and Celestia is the first application where I have such issues. The other ones are working fine without variables and stuff.

ColinKinloch commented:
SDL doesn't have many GUI features, it's just a simple multi platform way to get graphics on screen.

I can't see anything unusual in your locale. I assume the locale issue is upstream somewhere.

As for the addons could you try `~/.var/app/space.celestiaproject.Celestia/data/Celestia/`

RacerBG commented:
Adding add-ons this way doesn't work at all. From what I remember they were installed on Windows in the Celestia root directory in extras folder or something but here this is not a solution.

hfiguiere commented:
> I just gave it a test with `--nosocket=wayland`, it seems to work fine my end under XWayland.
> Is there a way to get this application to print a log?

The wayland permission is irrelevant for Xwayland. Xwayland is X11.

RacerBG commented:
> > And by the way why are you providing 3 different backends with a single package? The approach Freeciv is using seems better - they have different packages for every backend.
> 
> There should only be one. Qt or Gtk3, pick one.
> 
> And I don't condone the Freeciv approach because it is confusing too.

If you ask me the GTK one seems to be the best choice - it respects the system settings and it looks native even under KDE.

ColinKinloch commented:
The Qt frontend seems the most complete by a long way.

ColinKinloch commented:
bot, build space.celestiaproject.Celestia

I've added access to videos and pictures directory and added `~/Documents/Celestia/extras` to the configs extras directory list for addons.
I've also removed the metainfo and desktop files for the non Qt6 backends, they can still used `--command=` and they're don't add too much to the size of the package (561K for GTK, 388K for SDL).

RacerBG commented:
> added `~/Documents/Celestia/extras` to the configs extras directory list for addons.

Why in Documents? The initial path was fine `~/.var/app/space.celestiaproject.Celestia/data/Celestia/`

RacerBG commented:
bot, build space.celestiaproject.Celestia

RacerBG commented:
Loading add-onds from `~/.local/share/Celestia/extras` does not work probably because of lack of permissions.

I would also recommend adding `:create` to `xdg-videos` and `xdg-pictures` because the whole reason for their existence is to enable Celestia to create videos and pictures in these directories.

Other than that, everything else works now, thank you.

ColinKinloch commented:
The `:create` option creates the directory on start so probably unnecessary for `xdg-videos` etc.

I initially added the `xdg-documents` permission to test video recording, so I'll remove it in favour of `xdg-videos`.

I'm not sure whether I should leave those directories in the config file. It should probably be created on start in a user editable directory. Maybe I'll create a wrapper.

RacerBG commented:
> The `:create` option creates the directory on start so probably unnecessary for `xdg-videos` etc.
> 
> I initially added the `xdg-documents` permission to test video recording, so I'll remove it in favour of `xdg-videos`.
> 
> I'm not sure whether I should leave those directories in the config file. It should probably be created on start in a user editable directory. Maybe I'll create a wrapper.

Well I'm not that familiar with how everything works in flatpaks so it's up to you. As I said leaving only `~/.var/app/space.celestiaproject.Celestia` seems like the best decision and yes, this should be automatically created if possible. :smiley: 

Another question, where is the search functionality in the Qt version? On GTK pressing Ctrl+F did the job.

ColinKinloch commented:
Pressing Enter does a comparable thing on all frontends

RacerBG commented:
> Pressing Enter does a comparable thing on all frontends

Thank you!

RacerBG commented:
bot, build space.celestiaproject.Celestia

RacerBG commented:
bot, build space.celestiaproject.Celestia

RacerBG commented:
Very nice! I have no further recommendations. The only nitpick is the deafult directory for captured images - it opens `/app/share/celestia` instead it should open `~/Pictures` but this is not a big deal.

And one question about your recent PR in the main Celestia repo: Is it going to supersede your patch which you created here?

You have a green light from me, very good work so far and thank you for your cooperation! :+1:

hfiguiere commented:
> Very nice! I have no further recommendations. The only nitpick is the deafult directory for "Grab Image" - it opens `/app/share/celestia` instead it should open `~/Pictures` but this is not a big deal.

It is a big deal. `/app/share` is readonly and inaccessible.

ColinKinloch commented:
The "Grab Image" option brings up a file picker, so not too bad.
Still, it would be better to use portals, does Qt6 have an analogue to `GtkFileChooserNative`?

RacerBG commented:
Unfortunately the same can be said about "Open Script..."

"Capture Video" is fine.

RacerBG commented:
bot, build space.celestiaproject.Celestia

RacerBG commented:
bot, build space.celestiaproject.Celestia

RacerBG commented:
Everything opens correctly now. I was wrong about "Open Script..." because it's function is dfferent from the two options mentioned above.

:+1:

RacerBG commented:
bot, build space.celestiaproject.Celestia

RacerBG commented:
@ColinKinloch Please, remove the `LC_ALL` variable. It was my fault that Celestia wasn't working on my machine. Creating /etc/locale.conf with my choices of locale (Bulgarian and English) solved the issue.

ColinKinloch commented:
I've added a patch to make setting `LC_ALL` unnecessary.
The error was:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  locale::facet::_S_create_c_locale name not valid
```
Caused here: https://github.com/CelestiaProject/Celestia/blob/35878a35f1982b381683cc46e26c4c61011b9d94/src/celestia/celestiacore.h#L429 by `std::locale("")` throwing.
I think my patch could be improved by calling `std::locale::global` https://en.cppreference.com/w/cpp/locale/locale/locale though this may not change anything.

RacerBG commented:
bot, build space.celestiaproject.Celestia

ColinKinloch commented:
bot, build space.celestiaproject.Celestia

RacerBG commented:
bot, build space.celestiaproject.Celestia

ColinKinloch commented:
bot, build space.celestiaproject.Celestia

RacerBG commented:
bot, build space.celestiaproject.Celestia

RacerBG commented:
bot, build space.celestiaproject.Celestia

RacerBG commented:
bot, build space.celestiaproject.Celestia

RacerBG commented:
I'm almost certain that this is from upstream but it's worth mentioning that with this commit the planets are bugged...

file:///home/georgi/Pictures/saturn.png

Everything else works as expected.

ColinKinloch commented:
What do you mean by integrates with the system seamlessly?

RacerBG commented:
> What do you mean by integrates with the system seamlessly?

Now it respects the visual appearance from the system settings in KDE at least.

RacerBG commented:
@ColinKinloch Your PR's were recently merged in Celestia master, could you update the flatpak package in order to give it another round of testing?

ColinKinloch commented:
bot, build space.celestiaproject.Celestia

RacerBG commented:
Works as expected, the planet bug is gone too (which was indeed caused from upstream).

Anything else which needs to be tested?

RacerBG commented:
I'm curious to know what is the status of this PR and why it's still a draft?

ColinKinloch commented:
I initially wrote the manifest to build the master branch using cmake assuming I could just apply it to stable when ready, however the latest stable uses autotools.
I did attempt to compile stable 1.6.4 at one point but it seemed like it wasn't going to be trivial so I decided to wait for a stable 1.7.0.

RacerBG commented:
> I initially wrote the manifest to build the master branch using cmake assuming I could just apply it to stable when ready, however the latest stable uses autotools. I did attempt to compile stable 1.6.4 at one point but it seemed like it wasn't going to be trivial so I decided to wait for a stable 1.7.0.

I see, thanks for the information however it's uncertain when this will happen. I'm literally waiting for version 1.7.0 for the last 6-7 years or so...

ColinKinloch commented:
bot, build space.celestiaproject.Celestia
